### ISSUE TITLE
Mod Incompatibility with FarSight

### ISSUE THREAD

**Issue by FireController1847** [Zensonaton/MinecraftClientGobrrr#6]
Howdy! I was just thinking of creating a mod that does just this, and I came across yours! I cannot express how excited I got when I discovered it already exists (do you have a dono link? I wanna give you $5 lol).

Unfortunately, I'm having an issue where my game crashes with "Cannot invoke "farsight.config.Configuration.getCommonConfig()" because "farsight.FarsightMod.config" is null". I've attached the full crash report. It seems the mod has a conflict with farsight, though I'm not entirely sure why. Once I disabled farsight, the game launches fine.

Notably, another mod I have also conflicts, though this one would be more obvious, I figured I'd make you aware of it at the very least. It's a mod called [Window Title Changer](https://www.curseforge.com/minecraft/mc-mods/window-title-changer). Again, this is a very obvious conflict so I wouldn't *expect* anything to come out of this.

[crash-2023-06-03_01.38.52-client.txt](https://github.com/Zensonaton/MinecraftClientGobrrr/files/11642038/crash-2023-06-03_01.38.52-client.txt)

**Comment by Zensonaton** [comment]
Hey-ya! (No, I don't, and even if I did, then I wouldn't have accepted that.)

Both Bobby and Farsight are causing the exact same crash; both of those mods are trying to read the game configuration (to get the render distance setting value), however, my mod is doing some wacky magic in the background, and this is what causes those crashes. I was trying to fix that, but I couldn't find the 'proper' solution, unfortunately, but I'll try once again.

Window Title Changer case is pretty much interesting; yes, it is obvious that those would conflict, but I think that there *is* a way to make those compatible, I'll look into that.

I'm really grateful for your effortful (?) report/issue. I'll tag you here once I'll get something working.

**Comment by Zensonaton** [comment]
So, I couldn't implement a fix for Farsight/Bobby because it would break Iris mod, which is more evil-ly than compared to the Bobby/Farsight problem.

I've made [an issue #2020](https://github.com/IrisShaders/Iris/issues/2020), and we have to wait for the fix on Iris side.

**Comment by Zensonaton** [comment]
Unfortunately, the Window Title Changer crash is the same crash as with the Bobby or Farsight; this mod is trying to load the game configuration before its even loaded.

*sigh* it looks like I really have to find some kind of proper solution.

**Comment by suerion** [comment]
Hey, i had added an Issue on the farsight also, would link it here https://github.com/someaddons/farsight/issues/55

**Comment by warior456** [comment]
https://github.com/someaddons/farsight/issues/55#issuecomment-1595876433

**Comment by suerion** [comment]
> [someaddons/farsight#55 (comment)](https://github.com/someaddons/farsight/issues/55#issuecomment-1595876433)

is not fixed, he had only added the incompatibility for the mod

**Comment by Zensonaton** [comment]
> is not fixed, he had only added the incompatibility for the mod

True. The fun thing is that I've [did the same thing in my mod](https://github.com/Zensonaton/MinecraftClientGobrrr/blob/main/src/main/resources/fabric.mod.json#L37C9-L41).

@someaddons, "fix in 3.5" isn't the same thing as "I've marked the conflicting mod as `breaks` in Fabric mod metadata file".

**Comment by suerion** [comment]
Okey, 

Funfacts xD

**Comment by Zensonaton** [comment]
The issue preventing me from fixing the crash is the requirement for some specific adjustments that involve a bit of magic.

Let me explain. Both Farsight and Bobby are performing the exact same tasks, so let's take Bobby as an example.

Here's a simplified version of how Minecraft is loaded with the Bobby mod:
1. Load the mod Mixins.
2. Load other mods.
3. Load game options.
   + (Injected Bobby mod code via Mixin): Read game options during their loading and set `BobbyModClass.special_private_value = ...`.
4. Create the game window.

And here's how the game is loaded with both Bobby and my mod:
1. Load the mod Mixins.
2. [My mod] Load game options.
   + (Injected Bobby mod code via Mixin): Read game options during their loading and use `BobbyModClass.special_private_value = ...` (!).
3. [My mod] Create the game window.
4. Load other mods (!).

The problem arises because setting `BobbyModClass.special_private_value = ...` cannot be done before the mod is actually loaded by the Fabric mod loader. This is why the game crashes when using these mods.

I believe the only way to resolve this issue is to ask Bobby developers to implement a check that ensures their mod is loaded before attempting to set `BobbyModClass.special_private_value = ...`.
